# Практическая работа №1: Исследование алгоритмов формирования аддитивных цепочек

Выполнил студент гр. 0304 Мохаммед Али Васим, вариант 41.

## Цель работы 

Формирование представления о аддитивных цепочках, выработать умение составлять и применять алгоритмы для нахождения минимальных
аддитивных цепочек для заданного числа, привить навык использования систем компьютерной математики для реализации алгоритмов.

## Основные теоретические положения
**Бинарный метод** - метод быстрого возведения числа в натуральную степень $n$.  
Алгоритм:  
1) Записать число $n$ в бинарном виде  
2) Отбросить старший бит  
3) Произвести замену: если бит равен единице, то заменить его на SX иначе заменить на S  
4) Выполнить вычисление, где S - возведение в квадрат, а X - умножение на $x$   
**Метод множителей** - метод быстрого возведения числа в натуральную степень n.  
Алгортим:  
1) Представляем $n = p*q$, где $p$ - наименьший простой множитель $n$, $q > 1$. Таким образом $x^n$ можно найти, вычислив $x^p$ и возведя эту величину в степень $q$.  
2) Если $n$ - простое, то можно сначала вычислить $x^{n-1}$ и умножить его на x  
3) При $n = 1$ получим $x^n$ безо всяких вычислений.  
**Аддитивной цепочкой** для натурального числа $n$ называется последовательность натуральных чисел $a_0 = 1,a_1,a_2,...,a_m = n$, где каждый элемент последовательности равен сумме каких-либо двух предыдущих:
\\[ a_i = a_j + a_k, \; k \leq j \leq i, \; i = 1,2,...,r \\]  
**Алгоритм Брауэра** - метод нахождения аддитивной цепочки для натурального числа $n$.  
Алгоритм:  
Для натурального числа $n$ при заданном начальном числе $k$ можно построить цепочку Брауэра с помощью рекурсивной формулы:
$\begin{equation*}
 B_k(n) = 
     \begin{cases}
       1,2,3,...,2^k - 1 &\text{if $n < 2^k$}\\
           B_k(q),2*q,4*q,8*q,...,2^k*q,n &\text{if $n >= 2^k,  q = [\frac{n}{2^k}]$}
     \end{cases}
    \end{equation*}$  
1) Задаётся фиксированный параметр $k$ для рассматриваемого числа $n$. Вычисляются "вспомогательные числа":
$d = 2^k, q_1 = [\frac{n}{d}], r_1 = n \; mod \; d => n = q_1*d + r_1$  
$q_2 = [\frac{q1}{d}],r_2 = q_1 \; mod \; d => q_1 = q_2*d + r_2$  
2) Процедура продолжается до тех пор, пока не появится такое $q_s <  d => q_{s-1} = q_s*d + r_s$  
3) Таким образом $n$ имеет вид:
    \\[n = d^s*q_s + d^{s-1}*r_s + d^{s-2}*r_{s-1} + ... + \; r_1 \\]  
    **Алгоритм Яо** - метод нахождения аддитивной цепочки для натурального числа $n$.
    Алгоритм:
    Задаём некоторое целое $k >= 2$ и число $n$ раскладывается в $2^k$-й системе счисления:
    \\[ n = \sum\limits_{i=0}^j a_i*2^{i*k} \; a_j \neq 0 \\]
    Введём функцию d: 
    \\[d(z) = \sum_{i:a_i = z} 2^{i*k}\\]  
    1) Базовая последовательность:
    \\[1,2,4,8,...,2^{\lambda(n)}, \; где \; \lambda(n) - уменьшенная \; на \; единицу \; длина \; бинарной \; записи \; числа \; n \\]  
    2) Вычисление $d(z)$ для всех $z \in \{1,2,3,...,2^k-1\}, \; d(z) \neq 0$  
    3) Вычисление $z*d(z)$ для всех $z$  
    4) В конечном итоге, $n$ представляет собой разложение вида:
    \\[ n = \sum\limits_{z = 1}^{2^{k-1}} z*d(z) \\]  
    **Звёздной цепочкой** называется аддитивная цепочка включающая в себя только звёздные шаги.  
    Пара $(j,k), 0 \leq k \leq j < i \;$ называется  **шагом**  $i $. Тогда при $j = i-1$ пара называется **звёздным шагом**.

**Алгоритм дробления вектора индексов** - метод нахождения минимальной звёздной цепочки для натурального числа $n$.  
Алгоритм:  
1) Задаётся начальный вектор $r = \{1,2,3,...,m \}$ по которому строится начальная цепочка $a = \{a_1 = 1,a_2,a_3,...,a_{m+1}\}$  
2) Если $a_{m+1} = n$, то алгоритм завершается. В противном случае вектор r делится на две части: изменяемую и неизменяемую.  
3) Находим $a_{min}$ и $a_{max}$. Если $n \in [a_{min},a_{max}]$, то изменяемая часть вектора уменьшается на единицу по самому старшей позиции.  
4) Если цепочка так и не была найдена по всем возможным переборам изменяемой части вектора вплоть до $p = {1,1,...,1}$, то изменяемая часть принимает первоначальное значение, а неизменяемая уменьшается на единицу и процесс повторяется снова.  
5) Если были перебраны все варианты обоих частей вектора, то вектор $r$ увеличивается на единицу и принимает значение $r = {1,2,...,m+1}$.  
6) Алгоритм продолжается, пока не будет найдена цепочка.

## Выполнение работы
1) Были выполнены вычисления для $x^n$ бинарным методом и методом множителей для значений $n = 55, 139$:  
* **n = 55**  
   * *Двоичный метод "SX"* :
    \\[ 55 = 110111_2 \\]
    \\[ 10111_2 = SXSSXSXSX \\]
    \\[ SXSSXSXSX \longrightarrow \{x, x^2,x^3,x^6,x^{12},x^{13},x^{26},x^{27},x^{54}, x^{55} \}\\]  
   * *Метод множителей*:
    \\[ 55 = p*q = 5*11\\]
    \\[ y = x^5, \; y^{11} = (y^{10})*y = ((y^2)^2 * y)^2 * y \\]

* **n = 139**  
  *  *Двоичный метод "SX":*
    \\[ 139 = 10001011_2 \\]
    \\[ 0001011_2 = SSSSXSSXSX \\]
    \\[ SSSSXSSXSX \longrightarrow  \{x, x^2,x^4,x^8,x^{16},x^{17},x^{34},x^{68},x^{69},x^{138},x^{139}\} \\]  
  *  *Метод множителей:*
    \\[ 139 = p*q + 1 = 2*69 + 1 \\]
    \\[ y = x^2, \; y^{69} = (y^3)^{23} = ((((y^2*y)^2)^2)^2)^2*((y^2*y)^2)^2*(y^2*y)^2*(y^2*y)\\\\]  
  
    Количество операций представлено в табл. 1

|  n  | бинарный метод | Метод множителей |
|:---:|:--------------:|:----------------:|
|  55 |        9       |         8        |
| 139 |        10      |        15        |

\\[ \small{Таблица \; 1 - количество \; операций \; для \; бинарного \; метода \; и \; метода \; множителей} \\]

Исходя из результатов, приведённых в таблице, можно сказать, что метод множителей выполняет либо такое же количество операций, либо гораздо большее, когда наименьший простой множитель для числа $n$ равен двойке. В противном случае алгоритм способен выдавать лучшие результаты.

2) Был реализован алгоритм Брауэра для вычисления приближённых аддитивных цепочек при вариировании параметра $k$. Код представлен ниже:

In [1]:
import random

# Function for lambda 
def lbd(x):
    return floor(log(x, 2))

# Function for finding the Brauer chain of n with the base k 
def Brauer(n, k):
    if n < pow(2, k):
        return [_ for _ in range(1, (2 ^ k))]
    else:
        return Brauer(floor(n / 2 ^ k), k)  + [power(2, _)* floor(n / 2 ^ k) for _ in range(1, k+1)] + [n]
    
# Function for finding the minimal length of the chain
def min_length(n):
    return ceil(lbd(n) + (lbd(n) / (lbd(lbd(n))))) 

data   = [_ for _ in range(1, 2000)]
filter_data = []

# Part for randomn choice of numbers
#for i in range(2, 9, 3):
#    filter_data.append(int(str(random.choice(data))[0:i]))
    
# Part of manual choice of numbers    
sorted_filter_data = [15, 489, 1867]
data_k = [3, 5, 9]

for _ in sorted_filter_data:
    print(f"n = {_}")
    for k in data_k:
        print(f"Число k: {k} | Длина цепочки: {len(Brauer(_, k))} | Мин длина: {min_length(_)}")

n = 15
Число k: 3 | Длина цепочки: 11 | Мин длина: 6
Число k: 5 | Длина цепочки: 31 | Мин длина: 6
Число k: 9 | Длина цепочки: 511 | Мин длина: 6
n = 489
Число k: 3 | Длина цепочки: 15 | Мин длина: 11
Число k: 5 | Длина цепочки: 37 | Мин длина: 11
Число k: 9 | Длина цепочки: 511 | Мин длина: 11
n = 1867
Число k: 3 | Длина цепочки: 19 | Мин длина: 14
Число k: 5 | Длина цепочки: 43 | Мин длина: 14
Число k: 9 | Длина цепочки: 521 | Мин длина: 14


Результаты вычислений аддитивных цепочек при различных $n$ и параметре $k$ методом Брауэра представлены в табл. 2

|   n  | k | len  |min_len|
|:----:|:-:|:----:|:------|
| 15   | 3 |  11  |  6    |
| 15   | 5 |  31  |  6    |
| 15   | 9 |  511 |  6    | 
| 489  | 3 |  15  |  11   |
| 489  | 5 |  37  |  11   |
| 489  | 9 |  511 |  11   |
| 1867 | 3 |  19  |  14   |
| 1867 | 5 |  43  |  14   |
| 1867 | 9 |  521 |  14   |


\\[ \small{Таблица \; 2 - \; Таблица \; вычислений \; аддитивных \; цепочек \; методом \; Брауэра} \\]

Исходя из результатов таблицы, можно сделать вывод, что с увеличением параметра $k$ длина аддитивной цепочки увеличается. 

3)  Был реализован алгоритм дробления вектора индексов для нахождения минимальной звёздной цепочки натурального числа $n$. Код представлен ниже:

In [ ]:
from math import log
from math import floor
import time


# Function for building the chain with the index vector
def builder(vector):
    sequence = [1]
    for _ in range(0, len(vector)):
        sequence.append(sequence[_] + sequence[vector[_] - 1])
    return sequence


# Function for finding the previous vector
def previous(vector, original):
    tmp = vector[:]
    i = len(vector) - 1
    tmp[i] -= 1
    while tmp[i] == 0:
        tmp[i] = original[i]  # We go back to the original digit
        i -= 1                # We move down in the list
        if i == -1:
            break
        tmp[i] -= 1           # We reduce the next digit

    return tmp

# Function for lambda
def lambda_(n):
    return len(str(bin(n))[2:]) - 1

# Function for nu
def nu_(n):
    return str(bin(n))[2:].count('1')

# The general function
def min_sequence_processor(n):
    tic = time.perf_counter()
    
    m_down = lambda_(n) + floor(log(nu_(n), 2))
    m = m_down
    
    print(f"Мин длина цепочки: {m}")
    print("--- IN PROCESS ---")
    
    while 1:
        data, vector = min_sequence(n, m)
        if data == "end":
            m += 1
        else:
            tac = time.perf_counter()
            print("--- CHAIN FOUND ---")
            return data, vector, (tac - tic)


# The part of the function with the specific m
def min_sequence(n, m):

    # Find the divider index of the chain
    q = int(m / 2)

    # Divide the sequence's vector in two parts
    lv = list(range(1, q + 1));  rv = list(range(q + 1, m + 1))

    # Saving the originals sequences
    lv_0 = lv ; rv_0 = rv

    # Creating the data for the permutations
    # Iterators of the two lists
    while 1:
        # Creating the chain from the two vectors
        cs = builder(lv + rv)

        # If the latest element is n
        if cs[len(cs) - 1] == n:
            return cs, lv + rv

        # If n is in the chain
        elif (n >= cs[0]) and (n <= cs[len(cs) - 1]):
            while 1:
                if rv == [1] * len(rv) or lv == [1] * len(lv):   # If we reach all the possibles cases
                    return "end", [0]
                else:
                    rv = previous(rv, rv_0)     # If nothing special we change the right part
                    cs = builder(lv + rv)       # And we create the chain

                # If the latest element of the chain is n
                if cs[len(cs) - 1] == n:
                    return cs, lv + rv

        else:
            lv = previous(lv, lv_0)    # We change the left vector

            

# To choose randomly some values for the test

data   = [_ for _ in range(1, 4000)]
filter_data = []

for i in range(1, 5):
    filter_data.append(int(str(random.choice(data))[0:i]))
sorted_filter_data = sorted(filter_data)

# Part of manual choice of the numbers
sorted_filter_data = [1, 8, 27, 374, 1338, 2048, 3001, 4005]

for _ in sorted_filter_data:
    print(f"n: {_}")
    chain, vector, laps = min_sequence_processor(_)
    print(f"\nЦепочка: {chain}\nДлина цепочки: {len(chain)}\nВектор индексов: {vector}\nВремя нахождения: {laps:0.5f} s\n\n")

n: 1
Мин длина цепочки: 0
--- IN PROCESS ---
--- CHAIN FOUND ---

Цепочка: [1]
Длина цепочки: 1
Вектор индексов: []
Время нахождения: 0.00035 s


n: 8
Мин длина цепочки: 3
--- IN PROCESS ---
--- CHAIN FOUND ---

Цепочка: [1, 2, 4, 8]
Длина цепочки: 4
Вектор индексов: [1, 2, 3]
Время нахождения: 0.00011 s


n: 27
Мин длина цепочки: 6
--- IN PROCESS ---
--- CHAIN FOUND ---

Цепочка: [1, 2, 4, 8, 9, 18, 27]
Длина цепочки: 7
Вектор индексов: [1, 2, 3, 1, 5, 5]
Время нахождения: 0.00085 s


n: 374
Мин длина цепочки: 10
--- IN PROCESS ---
--- CHAIN FOUND ---

Цепочка: [1, 2, 4, 8, 16, 32, 64, 80, 82, 146, 292, 374]
Длина цепочки: 12
Вектор индексов: [1, 2, 3, 4, 5, 6, 5, 2, 7, 10, 9]
Время нахождения: 0.53240 s


n: 1338
Мин длина цепочки: 12
--- IN PROCESS ---


Результаты вычисления минимальной звёздной цепочки алгоритмом дробления вектора индексов представлены в табл. 3

|  n   |                            Цепочка                                      | len |        Время        |
|:----:|:------------------------------------------------------------------------|:---:|:-------------------:|
|  1   | 1                                                                       |  1  |   0.00012  seconds  |
|  8   | 1, 2, 4, 8                                                              |  4  |   0.00068  seconds  |
|  27  | 1, 2, 4, 8, 9, 18, 27                                                   |  7  |   0.00070  seconds  |
| 374  | 1, 2, 4, 8, 16, 32, 64, 80, 82, 146, 292, 374                           |  12 |   0.48501  seconds  |
| 1338 | 1, 2, 4, 8, 16, 32, 64, 96, 98, 194, 388, 776, 874                      |  14 |   7.77496  seconds  |
| 2048 | 1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 1024, 2048                       |  12 |   0.00007  seconds  |
| 3001 | 1, 2, 4, 8, 16, 32, 64, 128, 256, 272, 273, 546, 1092, 1364, 2728, 3001 |  16 |   375.8771 seconds  |
| 4005 | 1, 2, 4, 8, 16, 32, 64, 128, 256, 512, 768, 800, 801, 1602, 3204, 4005  |  16 |   186.5968 seconds  |

\\[ \small{Таблица \; 3 - \; Таблица \; вычислений \; минимальных \; звёздных \; цепочек}  \\]

Исходя из результатов таблицы можно сделать вывод о том, что для чисел $n$ кратных степеням двойки алгоритм находит цепочку достаточно быстро, в то время как поиск звёздных цепочек для остальных чисел занимает довольно длительное время. Это связано с тем, что алгоритм основан на переборе всех возможных вариантов вектора, по которому строится цепочка.

4) Проверим гипотезу Шольца-Брауэра для звёздных цепочек на алгоритме дробения индекса векторов:

In [ ]:
def Cholles_Brauer_testing():
    n = list(range(1, 13))
    v1 = list(map(lambda x: x - 1, n))            # v1 = n - 1 for n in [1, 12]
    v2 = list(map(lambda x: pow(2, x) - 1, n))    # v2 = 2^n - 1 for n in [1, 12]

    for _ in n:
        left,  lv, t1 = min_sequence_processor(v2[_ - 1])
        right, rv, t2 = min_sequence_processor(n[_ - 1])
        nmo = v1[_ - 1]

        print(f"I*(2^n - 1) = {len(left)} | I*(n) = {len(right)} | n = {n[_-1]} | Relation: {len(left) >= len(right) + nmo}")


Cholles_Brauer_testing()

Результаты проверки гипотезы Шольца-Брауэра представлены в табл. 4

| $$l^{*}(2^n-1)$$ | $$l^{*}(n)$$ | $$n$$ | $$l^{*}(2^{n}-1) \leq l^{*}(n) + n - 1$$ |
|:----------------:|:------------:|:-----:|:----------------------------------------:|
|         1        |       1      |   1   |                   True                   |
|         3        |       2      |   2   |                   True                   |
|         5        |       3      |   3   |                   True                   |
|         6        |       3      |   4   |                   True                   |
|         8        |       4      |   5   |                   True                   |
|         9        |       4      |   6   |                   True                   |
|        11        |       5      |   7   |                   True                   |
|        11        |       4      |   8   |                   True                   |
|        13        |       5      |   9   |                   True                   |
|        14        |       5      |   10  |                   True                   |
|        16        |       6      |   11  |                   True                   |
|        16        |       5      |   12  |                   True                   |

\\[ \small{Таблица \; 4 - проверка \; гипотезы \; Шольца-Брауэра}    \\]

Из результатов таблицы видно, что гипотеза Шольца-Брауэра верна для всех $n \in [1,12]$

## Выводы
В ходе выполнения практической работы были реализованы алгоритмы для создания аддитивных цепочек алгоритмом Брауэра и алгоритмом дробления вектора индексов. Исследования показали, что алгоритм Брауэра строит не самую минимальную цепочку для заданного числа $n$, а лишь какуе-то. Алгоритм дробления вектора индексов же строит минимальную звёздную цепочку, но затрачивает на это достаточно большое количество времени ввиду того, что является переборным. Гипотеза Шольца-Брауэра успешно подтверждена на алгоритме дробления вектора индексов.